File: TarsEngine.FSharp.Core/TreeOfThought/ThoughtNodeWrapper.fs
namespace TarsEngine.FSharp.Core.TreeOfThought

open System.Collections.Generic

/// <summary>
/// Implementation of the IThoughtNode interface that wraps a ThoughtNode.ThoughtNode.
/// </summary>
type ThoughtNodeWrapper(node: ThoughtNode.ThoughtNode) =
    /// <summary>
    /// Gets the thought content.
    /// </summary>
    member _.Thought = node.Thought
    
    /// <summary>
    /// Gets the child nodes.
    /// </summary>
    member _.Children =
        node.Children
        |> List.map (fun child -> ThoughtNodeWrapper(child) :> IThoughtNode)
        |> List.toArray
        |> (fun arr -> arr :> IReadOnlyList<IThoughtNode>)
    
    /// <summary>
    /// Gets the evaluation metrics.
    /// </summary>
    member _.Evaluation = node.Evaluation
    
    /// <summary>
    /// Gets a value indicating whether the node has been pruned.
    /// </summary>
    member _.Pruned = node.Pruned
    
    /// <summary>
    /// Gets the metadata.
    /// </summary>
    member _.Metadata =
        node.Metadata
        |> Map.toSeq
        |> Seq.map (fun (k, v) -> KeyValuePair<string, obj>(k, v))
        |> Dictionary<string, obj>
        |> (fun dict -> dict :> IReadOnlyDictionary<string, obj>)
    
    /// <summary>
    /// Gets the score of the node.
    /// </summary>
    member _.Score = ThoughtNode.getScore node
    
    /// <summary>
    /// Gets the underlying ThoughtNode.
    /// </summary>
    member _.Node = node
    
    interface IThoughtNode with
        member this.Thought = this.Thought
        member this.Children = this.Children
        member this.Evaluation = this.Evaluation
        member this.Pruned = this.Pruned
        member this.Metadata = this.Metadata
        member this.Score = this.Score

/// <summary>
/// Module containing functions for working with ThoughtNodeWrapper.
/// </summary>
module ThoughtNodeWrapper =
    /// <summary>
    /// Creates a new ThoughtNodeWrapper from a ThoughtNode.
    /// </summary>
    /// <param name="node">The ThoughtNode.</param>
    /// <returns>A new ThoughtNodeWrapper.</returns>
    let fromThoughtNode (node: ThoughtNode.ThoughtNode) =
        ThoughtNodeWrapper(node)
    
    /// <summary>
    /// Gets the underlying ThoughtNode from a ThoughtNodeWrapper.
    /// </summary>
    /// <param name="node">The ThoughtNodeWrapper.</param>
    /// <returns>The underlying ThoughtNode.</returns>
    let toThoughtNode (node: ThoughtNodeWrapper) =
        node.Node
    
    /// <summary>
    /// Gets the underlying ThoughtNode from an IThoughtNode.
    /// </summary>
    /// <param name="node">The IThoughtNode.</param>
    /// <returns>The underlying ThoughtNode.</returns>
    let toThoughtNodeFromInterface (node: IThoughtNode) =
        match node with
        | :? ThoughtNodeWrapper as wrapper -> wrapper.Node
        | _ -> failwith "Not a ThoughtNodeWrapper"

File: v1/parked_csharp/TarsEngine.FSharp.Adapters/TreeOfThought/ThoughtNodeWrapper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TarsEngine.Services.Abstractions.TreeOfThought;

namespace TarsEngine.FSharp.Adapters.TreeOfThought
{
    /// <summary>
    /// Wrapper for ThoughtNodeAdapter that implements IThoughtNode.
    /// </summary>
    public class ThoughtNodeWrapper : IThoughtNode
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="ThoughtNodeWrapper"/> class.
        /// </summary>
        /// <param name="node">The thought node adapter.</param>
        public ThoughtNodeWrapper(ThoughtNodeAdapter node)
        {
            Node = node ?? throw new ArgumentNullException(nameof(node));
        }

        /// <summary>
        /// Gets the thought node adapter.
        /// </summary>
        public ThoughtNodeAdapter Node { get; }

        /// <summary>
        /// Gets the thought content.
        /// </summary>
        public string Thought => Node.Thought;

        /// <summary>
        /// Gets the child nodes.
        /// </summary>
        public IReadOnlyList<IThoughtNode> Children => 
            Node.Children.Select(child => new ThoughtNodeWrapper(child)).ToList();

        /// <summary>
        /// Gets the evaluation metrics.
        /// </summary>
        public EvaluationMetrics? Evaluation => 
            Node.Evaluation != null 
                ? new EvaluationMetrics(
                    Node.Evaluation.Correctness,
                    Node.Evaluation.Efficiency,
                    Node.Evaluation.Robustness,
                    Node.Evaluation.Maintainability,
                    Node.Evaluation.Overall)
                : null;

        /// <summary>
        /// Gets a value indicating whether the node has been pruned.
        /// </summary>
        public bool Pruned => Node.Pruned;

        /// <summary>
        /// Gets the metadata.
        /// </summary>
        public IReadOnlyDictionary<string, object> Metadata => Node.Metadata;

        /// <summary>
        /// Gets the score of the node.
        /// </summary>
        public double Score => Node.GetScore();
    }
}
